src/abi_arm.cpp

```diff
@@ -0,0 +1,102 @@
+    if (jl_is_structtype(ty)) {
+        members = jl_datatype_nfields(ty);
+        for (size_t i = 0; i < members; ++i) {
+            jl_value_t *fty = jl_field_type((jl_datatype_t*)ty,i);
+            if (!base)
+                base = fty;
+            else if (base != fty)
+                return false;
+        }
+        return true;
+    }
+
+    return false;
+}
+
+bool use_sret(AbiState *state,jl_value_t *ty)
+{
+    if (!jl_is_datatype(ty) || jl_is_abstracttype(ty) || jl_is_cpointer_type(ty) || jl_is_array_type(ty))
+        return false;
+    size_t size = jl_datatype_size(ty);
+
+    // A double-word sized Fundamental Data Type (e.g., long long, double and
+    // 64-bit containerized vectors) is returned in r0 and r1.
+    if (jl_is_bitstype(ty) && size <= 8)
+        return false;
+
+    // A Composite Type not larger than 4 bytes is returned in r0. The format is
+    // as if the result had been stored in memory at a word-aligned address and
+    // then loaded into r0 with an LDR instruction. Any bits in r0 that lie
+    // outside the bounds of the result have unspecified values.
+    if (jl_is_structtype(ty) && size <= 4)
+        return false;
+
+    // For the VFP the following argument types are VFP CPRCs:
+    // - ...
+    // - A Homogeneous Aggregate with a Base Type of a single- or
+    //   double-precision floating-point type with one to four Elements.
+    // Any result whose type would satisfy the conditions for a VFP CPRC is
+    // returned in the appropriate number of consecutive VFP registers
+    const jl_value_t *base = nullptr;
+    size_t members = 0;
+    if (isHomogeneousAggregate(ty, base, members) && members <= 4 &&
+        (base == (jl_value_t*)jl_float64_type ||
+         base == (jl_value_t*)jl_float32_type))
+        return false;
+
+    return true;
+}
+
+void needPassByRef(AbiState *state,jl_value_t *ty, bool *byRef, bool *inReg)
+{
+    return;
+}
+
+Type *preferred_llvm_type(jl_value_t *ty, bool isret)
+{
+    // TODO: If the argument is an integral Fundamental Data Type that is
+    // smaller than a word, then it is zero- or sign-extended to a full word and
+    // its size is set to 4 bytes. If the argument is a Half-precision Floating
+    // Point Type it is converted to Single Precision.
+
+    if (jl_is_structtype(ty)) {
+        const jl_value_t *base = nullptr;
+        size_t members = 0;
+        if (!isHomogeneousAggregate(ty, base, members)) {
+            // Align heterogeneous aggregates
+            // TODO: can this be done using the 'align' attribute?
+            size_t max_size = 0;
+            for (size_t i = 0; i < members; ++i) {
+                jl_value_t *fty = jl_field_type((jl_datatype_t*)ty,i);
+                size_t size = jl_datatype_size(fty);
+                max_size = size>max_size ? size : max_size;
+            }
+            Type* T = Type::getIntNTy(getGlobalContext(), max_size*8);
+            return ArrayType::get(T, members);
```

Comment:
looking at clang ast, i think this can be implemented as:
```c
size_t align = jl_datatype_alignment(ty);
size_t size = jl_datatype_size(ty);
Type* T = Type::getIntNTy(getGlobalContext(), align*8);
return ArrayType::get(T, (size + align - 1) / align);
```
(the caller of this function will do the same cast-through-memory as clang to make this work)